src/atama/receiver/node/receiver_node.cpp

```diff
@@ -0,0 +1,113 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "atama/receiver/node/receiver_node.hpp"
+#include "kansei_interfaces/msg/orientation.hpp"
+#include "ninshiki_interfaces/msg/detected_object.hpp"
+#include "ninshiki_interfaces/msg/detected_objects.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "tachimawari_interfaces/srv/get_joints.hpp"
+
+using namespace std::chrono_literals;
+
+namespace atama::receiver
+{
+
+ReceiverNode::ReceiverNode(rclcpp::Node::SharedPtr node, std::shared_ptr<atama::head::Head> head)
+: node(node), head(head)
+{
+  using ninshiki_interfaces::msg::DetectedObject;
+
+  get_joints_client = node->create_client<GetJoints>("/joint/get_joints");
+
+  get_orientation_subsciber = node->create_subscription<Orientation>(
+    "measurement/orientation", 10,
+    [this](const Orientation::SharedPtr message) {
+      // this->head->set_yaw(message->orientation[2]);
+    }
+  );
+
+  get_detection_result_subsciber =
+    node->create_subscription<DetectedObjects>(
+    "ninshiki_py/detection", 10,
+    [this](const DetectedObjects::SharedPtr message) {
+      std::vector<DetectedObject> temp_detection_result;
+
+      for (const auto & detected_object : message->detected_objects) {
+        temp_detection_result.push_back(detected_object);
+      }
+
+      this->head->detection_result = temp_detection_result;
+    }
+    );
+}
+
+bool ReceiverNode::get_joints_data()
+{
+  using tachimawari::joint::Joint;
+  using tachimawari::joint::JointId;
+
+  while (!get_joints_client->wait_for_service(1s)) {
+    if (rclcpp::ok()) {
+      // service not available, waiting again...
+    } else {
+      // Interrupted while waiting for the service. Exiting.
+      return false;
+    }
+  }
+  auto result = get_joints_client->async_send_request(
+    std::make_shared<tachimawari_interfaces::srv::GetJoints::Request>());
+
+  if (rclcpp::spin_until_future_complete(node, result) ==
+    rclcpp::FutureReturnCode::SUCCESS)
+  {
+    std::vector<Joint> temp_joints;
+    std::set<uint8_t> joints_id;
+
+    for (std::string id : {"neck_yaw", "neck_pitch"}) {
+      joints_id.insert(JointId::by_name.find(id)->second);
```

Comment:
there's a possibility that `JointId::by_name.find(id)` will return `JointId::by_name.end()` here

Reply:
What the effect if `JointId::by_name.find(id)` return `JointId::by_name.end()`? Should I add an if guard here?

Reply:
> What the effect if `JointId::by_name.find(id)` return `JointId::by_name.end()`? Should I add an if guard here?

i guess `JointId::by_name.end()->second` will be invalid? idk though whether it could cause crash or not.

Reply:
![image](https://user-images.githubusercontent.com/10202888/157584474-2c37761c-825b-4b43-8acc-e84a2a2eaca5.png)
accessing `second` could result to undefined behavior

Reply:
ok I will add if guard here